functions/venue.js

```diff
@@ -88,6 +88,9 @@ const createVenueData = (data, context) => {
     requiresDateOfBirth: data.requiresDateOfBirth || false,
     showRadio: data.showRadio || false,
     radioStations: data.radioStations ? [data.radioStations] : [],
+    ...(data.mapBackgroundImageUrl && {
```

Comment:
What are the downsides of doing it like this?

```mapBackgroundImageUrl: data.mapBackgroundImageUrl,```

Reply:
You can not have undefined in the firebase. And since some templates don't have mapBackgroundImageUrl, I added this conditional to make sure this is uploaded to the firebase only when it exists. We can use a default value but then we an extra field in the database which is is not being used for some templates.

Reply:
To explain in more in details:

If we use `mapBackgroundImageUrl: data.mapBackgroundImageUrl` and then create a jazzbar, the function will fail because `mapBackgroundImageUrl` will be `undefined`

If we use `mapBackgroundImageUrl: data.mapBackgroundImageUrl ?? "` and then create a jazzbar, the function will succeed but there will be a field `mapBackgroundImageUrl: ""` in the database that will never be used.

So that leaves us with 2 options: 
1. Add `mapBackgroundImageUrl: data.mapBackgroundImageUrl` to the switch cases for all of the templates that have `mapBackgroundImageUrl` and then keep adding it in the future ones.
2. Add conditional `mapBackgroundImageUrl` that will add the field only when it exists.

Reply:
> You can not have `undefined` in the firebase.

> If we use `mapBackgroundImageUrl: data.mapBackgroundImageUrl` and then create a `jazzbar`, the function will fail because `mapBackgroundImageUrl` will be `undefined`

@denisdimitrov With regards to these comments, what specifically happens if we try and set a key that is `undefined`? Does the Firebase SDK throw an error, or it just doesn't set that key, or?

If there is never a valid case where this data structure can contain `undefined` keys, then we could perhaps add our own more generic 'filter' pattern, where we treat `undefined` keys as 'do not set', and then filter them out from this structure before passing them to the Firebase SDK. There are many ways we could do that if we wanted to, but here's a one (and others in that thread for doing it with native JavaScript):

- https://stackoverflow.com/questions/25421233/javascript-removing-undefined-fields-from-an-object/37836669#37836669

---

@denisdimitrov I know you already said that your intent was to not have to make too many future changes as we add new templates, but I'm curious, which templates currently use `mapBackgroundImageUrl`? (guessing `partymap` and `themecamp` (deprecated))

If it's only for these `map` templates, I don't know that there is going to be a lot of 'future churn'. I believe we decided a while back that going forward we're going to 'focus in' and only have the main `partymap`, and then the smaller selection of 'other venues'; so I don't expect we'll be adding many/any new `map` venues.

I kind of feel like the way it's been implemented in this diff is probably 'worse' than how it was originally implemented before the suggestions in https://github.com/sparkletown/sparkle/pull/1071#pullrequestreview-568118430 were implemented.

Reply:
>@denisdimitrov With regards to these comments, what specifically happens if we try and set a key that is undefined? Does the Firebase SDK throw an error, or it just doesn't set that key, or?


@0xdevalias Firebase throws an error and the request fails. That's why some of the values are being added with if statements conditionally.

>If there is never a valid case where this data structure can contain undefined keys, then we could perhaps add our own more generic 'filter' pattern, where we treat undefined keys as 'do not set', and then filter them out from this structure before passing them to the Firebase SDK. There are many ways we could do that if we wanted to, but here's a one (and others in that thread for doing it with native JavaScript):

@0xdevalias Imo this is an overkill, instead of adding undefined values and then filtering them out, why not just add the actual values conditionally? Just the way they are now, this way you skip the object iteration.

> I kind of feel like the way it's been implemented in this diff is probably 'worse' than how it was originally implemented before the suggestions in #1071 (review) were implemented.

@0xdevalias This diff and the previous one are identical, its just the syntax that is different. I'm conditionally adding the field to the object values instead of using if statement.


**These do the same thing:**
```jsx
const venueData = {}
if (data.mapBackgroundImageUrl) {
    venueData.mapBackgroundImageUrl = data.mapBackgroundImageUrl;
}
```

```jsx
const venueData = {
  ...(data.mapBackgroundImageUrl && {
    mapBackgroundImageUrl: data.mapBackgroundImageUrl,
  })
}
```